index.bs

```diff
@@ -931,6 +1019,231 @@ steps:
+as well as obtaining and changing the size of, a single file.
+
+The {{FileSystemSyncAccessHandle/read()}} and {{FileSystemSyncAccessHandle/write()}} methods are synchronous.
+This allows for higher performance for critical methods on contexts where asynchronous
+operations come with high overhead, e.g., WebAssembly.
+
+</div>
+
+<div algorithm>
+To <dfn>create a new FileSystemSyncAccessHandle</dfn> given a [=file entry=] |file|
+in a [=/Realm=] |realm|, perform the following steps:
+
+1. Let |handle| be a [=new=] {{FileSystemSyncAccessHandle}} in |realm|.
+1. Set |handle|.[=FileSystemSyncAccessHandle/[[file]]=] to |file|.
+1. Set |handle|.[=FileSystemSyncAccessHandle/[[state]]=] to "`open`".
+1. Return |handle|.
+
+</div>
+
+### The {{FileSystemSyncAccessHandle/read()}} method ### {#api-filesystemsyncaccesshandle-read}
+
+<div class="note domintro">
+  : |handle| . {{FileSystemSyncAccessHandle/read()|read}}(|buffer|)
+  : |handle| . {{FileSystemSyncAccessHandle/read()|read}}(|buffer|, { {{FileSystemReadWriteOptions/at}} })
+  :: Reads the contents of the file associated with |handle| into |buffer|, optionally at a given offset.
+</div>
+
+<div algorithm>
+The <dfn method for=FileSystemSyncAccessHandle>read(|buffer|, {{FileSystemReadWriteOptions}}: |options|)</dfn> method, when invoked, must run
+these steps:
+
+1. If [=this=].[=[[state]]=] is "`closed`", throw an {{InvalidStateError}}.
+1. Let |bufferSize| be |buffer|'s [=byte length=].
+1. Let |fileContents| be [=this=].[=[[file]]=]'s [=file entry/binary data=].
+1. Let |fileSize| be |fileContents|'s [=byte sequence/length=].
+1. Let |readStart| be |options|.{{FileSystemReadWriteOptions/at}}.
+1. If |readStart| is larger than |fileSize|, return 0.
+1. Let |readEnd| be |readStart| + (|bufferSize| &minus; 1).
+1. If |readEnd| is larger than |fileSize|, set |readEnd| to |fileSize|.
+1. Let |result| be |bufferSize|.
+1. Let |bytes| be a [=byte sequence=] containing the bytes from |readStart| to |readEnd| of |fileContents|.
+1. If the operations reading from |fileContents| in the previous steps failed:
+    1. If there were partial reads  and the number of modified bytes in |bytes|  is known,
+       set |result| to the number of modified bytes.
+    1. Else set |result| to 0.
+1. Let |arrayBuffer| be |buffer|'s [=underlying buffer=].
+1. [=write|Write=] |bytes| into |arrayBuffer|.
+1. Return |result|.
+
+</div>
+
+### The {{FileSystemSyncAccessHandle/write()}} method ### {#api-filesystemsyncaccesshandle-write}
+
+<div class="note domintro">
+  : |handle| . {{FileSystemSyncAccessHandle/write()|write}}(|buffer|)
+  : |handle| . {{FileSystemSyncAccessHandle/write()|write}}(|buffer|, { {{FileSystemReadWriteOptions/at}} })
+  :: Writes the content of |buffer| into the file associated with |handle|, optionally at a given offset.
+</div>
+
+// TODO(fivedots): Figure out how to properly check the available storage quota (in this method and others) by passing the right storage shelf.
```

Comment:
Any plans to resolve this todo in a manner that we can implement (without all of Storage API)?   What does chrome's current implementation do, and what does Safari do?

Reply:
The real missing part here is specifying what happens in quota cases for write() (as is done in [WritableFileStream]([quota](https://storage.spec.whatwg.org/#storage-quota))): 

> If the operations modifying stream.[[[buffer]]](https://fs.spec.whatwg.org/#filesystemwritablefilestream-buffer) in the previous steps failed due to exceeding the [storage quota](https://storage.spec.whatwg.org/#storage-quota), [reject](https://webidl.spec.whatwg.org/#reject) p with a [QuotaExceededError](https://webidl.spec.whatwg.org/#quotaexceedederror) and abort, leaving stream.[[[buffer]]](https://fs.spec.whatwg.org/#filesystemwritablefilestream-buffer) unmodified.

> Note: [Storage quota](https://storage.spec.whatwg.org/#storage-quota) only applies to files stored in the [origin private file system](https://fs.spec.whatwg.org/#origin-private-file-system). However this operation could still fail for other files, for example if the disk being written to runs out of disk space.

Reply:
(A storage shelf is essentially the storage space for an "origin" (really storage key). https://storage.spec.whatwg.org/#model goes into this in more detail.)

Reply:
Ok